packages/turbo-repository/rust/src/lib.rs

```diff
@@ -131,4 +135,56 @@ impl Workspace {
 
         Ok(map)
     }
+
+    #[napi]
+    pub async fn affected_packages(&self, files: Vec<String>) -> Result<Vec<Package>, Error> {
+        let mapper = ChangeMapper::new(&self.graph, vec![], vec![]);
+
+        let workspace_root = match AbsoluteSystemPath::new(&self.absolute_path) {
+            Ok(path) => path,
+            Err(e) => return Err(Error::from_reason(e.to_string())),
+        };
+
+        let hash_set_of_paths: HashSet<AnchoredSystemPathBuf> = files
+            .into_iter()
+            .filter_map(|s| {
+                let split = s.split(std::path::MAIN_SEPARATOR).collect::<Vec<&str>>();
+                let xx = workspace_root.join_components(&split);
+                match workspace_root.anchor(&xx) {
+                    Ok(path) => Some(path),
+                    Err(e) => None,
+                }
+            })
+            .collect();
+
+        let package_changes = match mapper.changed_packages(hash_set_of_paths, None) {
+            Ok(changes) => changes,
+            Err(e) => return Err(Error::from_reason(e.to_string())),
+        };
+
+        let packages = match package_changes {
+            PackageChanges::All => self
+                .graph
+                .workspaces()
+                .map(|(name, info)| WorkspacePackage {
+                    name: name.to_owned(),
+                    path: info.package_path().to_owned(),
+                })
+                .collect::<Vec<WorkspacePackage>>(),
+            PackageChanges::Some(packages) => packages.into_iter().collect(),
+        };
+
+        let mut serializable_packages: Vec<Package> = packages
+            .into_iter()
+            .map(|p| {
+                let package_path = workspace_root.resolve(&p.path);
+                Package::new(p.name.to_string(), &workspace_root, &package_path)
+            })
+            .filter(|p| p.name != "//")
+            .collect();
+
+        serializable_packages.sort();
```

Comment:
I'm not sure what this `Sort` is doing on a custom struct, but I want it to sort by the `name` field. Might need to implement the `Ord` trait manually?

Reply:
If I understand correctly you probably wanna use [this std function?]([url](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.sort_by_key))

Reply:
This is the comparison that's being used to sort (can be found by using [expand macro](https://rust-analyzer.github.io/manual.html#expand-macro-recursively) on the `#[derive(Ord)]` declaration:
```
impl $crate::cmp::Ord for Package {
    fn cmp(&self, other: &Self) -> $crate::cmp::Ordering {
        match (self, other) {
            (
                Package {
                    name: name_self,
                    absolute_path: absolute_path_self,
                    relative_path: relative_path_self,
                },
                Package {
                    name: name_other,
                    absolute_path: absolute_path_other,
                    relative_path: relative_path_other,
                },
            ) => match name_self.cmp(&name_other) {
                $crate::cmp::Ordering::Equal => {
                    match absolute_path_self.cmp(&absolute_path_other) {
                        $crate::cmp::Ordering::Equal => {
                            match relative_path_self.cmp(&relative_path_other) {
                                $crate::cmp::Ordering::Equal => $crate::cmp::Ordering::Equal,
                                c => return c,
                            }
                        }
                        c => return c,
                    }
                }
                c => return c,
            },
            _unused => $crate::cmp::Ordering::Equal,
        }
    }
}
```
If you just want to sort based on names, then you should look at [sort_by_key](https://doc.rust-lang.org/std/primitive.slice.html#method.sort_by_key)

Reply:
Thanks for the pointer, updated to `sort_by_key`